cli/up/build.go

```diff
@@ -47,38 +59,48 @@ func (cmd *up) buildImages(composeFile composeTypes.Config) (map[string]string,
 		return nil, errors.New("no docker client")
 	}
 
-	images := map[string]image{}
-	svcToImageName := map[string]string{}
-	var imageNames []string
+	var buildServices composeTypes.Services
 	for _, svc := range composeFile.Services {
-		if svc.Build == nil {
-			continue
+		if svc.Build != nil {
+			buildServices = append(buildServices, svc)
 		}
+	}
+
+	toPush := cmd.getServicesToPush(buildServices)
 
+	readyToPush, err := cmd.prePushBaseImages(toPush)
+	if err != nil {
+		log.WithError(err).Debug("Failed to pre-push the base images. " +
+			"Falling back to pushing full images.")
+	}
+
+	images := map[string]image{}
+	svcToImageName := map[string]string{}
+	for _, svc := range buildServices {
 		imageID, err := cmd.buildImage(*svc.Build, svc.Name)
 		if err != nil {
 			return nil, errors.WithContext(fmt.Sprintf("build %s", svc.Name), err)
 		}
 
-		imageName := fmt.Sprintf("%s/%s:%s", cmd.imageNamespace, svc.Name,
-			strings.TrimPrefix(imageID, "sha256:"))
+		imageName := getImageName(cmd.imageNamespace, svc.Name, imageID)
 		svcToImageName[svc.Name] = imageName
 		images[svc.Name] = image{imageID, imageName}
-		imageNames = append(imageNames, imageName)
 	}
 
-	// When pushing an image, we first check to see if the remote manifest
-	// already exists. This is more efficient than doing a full image push
-	// because we don't compare each individual layer.
-	pushedImages := getPushedImages(imageNames, cmd.auth.AuthToken)
-	for svc, img := range images {
-		if _, exists := pushedImages[img]; exists {
-			log.WithField("service", svc).Debug("Skipping push. Remote image already exists.")
-			continue
+	// We expect each service to be sent exactly once, and we expect the channel
+	// to be closed when all images have been signalled as ready to push.
+	for base := range readyToPush {
+		if base.imageName != "" {
+			err = cmd.pushBaseTag(base.imageName, base.service)
```

Comment:
A comment about why we need to re-push the base image would be helpful, docker's behavior is very... non-obvious.

Reply:
Also does the UI look reasonable now that you're doing two pushes for each service?

Reply:
> A comment about why we need to re-push the base image would be helpful, docker's behavior is very... non-obvious.

Huh, I remember writing a comment for this... Not sure what happened to it, will readd.

> Also does the UI look reasonable now that you're doing two pushes for each service?

Yes, it just shows the two pushes sequentially. It could be slightly confusing to have the extra push, but I think it shows the tag when pushing, which is relatively descriptive (<service>:base).